# Homework

## Q1. Install MLflow

In [1]:
!mlflow --version

mlflow, version 2.22.0


Answer: **2.22.0**

## Q2. Download and preprocess the data

1. Download the datasets

In [3]:
!wget https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2023-01.parquet -P ./dataset
!wget https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2023-02.parquet -P ./dataset
!wget https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2023-03.parquet -P ./dataset

--2025-05-22 00:47:51--  https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2023-01.parquet
Resolving d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)... 2600:9000:2795:8e00:b:20a5:b140:21, 2600:9000:2795:7000:b:20a5:b140:21, 2600:9000:2795:b200:b:20a5:b140:21, ...
Connecting to d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)|2600:9000:2795:8e00:b:20a5:b140:21|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1427002 (1.4M) [binary/octet-stream]
Saving to: ‘./dataset/green_tripdata_2023-01.parquet’

green_tripdata_2023 100%[===================>]   1.36M  1.30MB/s    in 1.0s    

2025-05-22 00:47:53 (1.30 MB/s) - ‘./dataset/green_tripdata_2023-01.parquet’ saved [1427002/1427002]

--2025-05-22 00:47:53--  https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2023-02.parquet
Resolving d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)... 2600:9000:2795:6a00:b:20a5:b140:21, 2600:9000:2795:7000:b:20a5:b140:21

2. Execute the command

In [4]:
!python preprocess_data.py --raw_data_path ./dataset --dest_path ./output

How many files were saved to `OUTPUT_FOLDER`?

In [5]:
!ls ./output

dv.pkl    test.pkl  train.pkl val.pkl


Answer: **4 files**

## Q3. Train a model with autolog

1. Modify the `train.py` script to enable autologging

In [ ]:
# Modified `train.py` to enable autologging in MLflow and use `with mlflow.start_run():` wrapper 
!cat train.py

2. Execute the script

In [8]:
!python train.py

2025/05/22 01:07:12 INFO mlflow.tracking.fluent: Autologging successfully enabled for sklearn.
2025/05/22 01:07:20 WARNING mlflow.utils.autologging_utils: Encountered unexpected error during sklearn autologging: When an mlflow-artifacts URI was supplied, the tracking URI must be a valid http or https URI, but it was currently set to file:///Users/nop/Documents/GitHub/mlops-zoomcamp-datatalks/02-experiment-tracking/mlruns. Perhaps you forgot to set the tracking URI to the running MLflow server. To set the tracking URI, use either of the following methods:
1. Set the MLFLOW_TRACKING_URI environment variable to the desired tracking URI. `export MLFLOW_TRACKING_URI=http://localhost:5000`
2. Set the tracking URI programmatically by calling `mlflow.set_tracking_uri`. `mlflow.set_tracking_uri('http://localhost:5000')`


3. Check `min_sample_split`

In [ ]:
from mlflow.tracking import MlflowClient

client = MlflowClient()

In [12]:
client.search_experiments()

[<Experiment: artifact_location='mlflow-artifacts:/0', creation_time=1747848584001, experiment_id='0', last_update_time=1747848584001, lifecycle_stage='active', name='Default', tags={}>]

In [14]:
client.search_runs(experiment_ids="0")

[<Run: data=<RunData: metrics={}, params={'bootstrap': 'True',
  'ccp_alpha': '0.0',
  'criterion': 'squared_error',
  'max_depth': '10',
  'max_features': '1.0',
  'max_leaf_nodes': 'None',
  'max_samples': 'None',
  'min_impurity_decrease': '0.0',
  'min_samples_leaf': '1',
  'min_samples_split': '2',
  'min_weight_fraction_leaf': '0.0',
  'monotonic_cst': 'None',
  'n_estimators': '100',
  'n_jobs': 'None',
  'oob_score': 'False',
  'random_state': '0',
  'verbose': '0',
  'warm_start': 'False'}, tags={'estimator_class': 'sklearn.ensemble._forest.RandomForestRegressor',
  'estimator_name': 'RandomForestRegressor',
  'mlflow.runName': 'bustling-deer-805',
  'mlflow.source.git.commit': '8694135104718c1a64b979de8a920ebd0223bafe',
  'mlflow.source.name': 'train.py',
  'mlflow.source.type': 'LOCAL',
  'mlflow.user': 'nop'}>, info=<RunInfo: artifact_uri='mlflow-artifacts:/0/15102ccad1b049118b503017f8d93929/artifacts', end_time=1747850840629, experiment_id='0', lifecycle_stage='active', ru

In [16]:
RUN_ID = "15102ccad1b049118b503017f8d93929"

run = client.get_run(run_id=RUN_ID)
run.data.params

{'bootstrap': 'True',
 'max_depth': '10',
 'max_samples': 'None',
 'min_weight_fraction_leaf': '0.0',
 'max_leaf_nodes': 'None',
 'min_samples_leaf': '1',
 'random_state': '0',
 'min_impurity_decrease': '0.0',
 'verbose': '0',
 'n_estimators': '100',
 'criterion': 'squared_error',
 'oob_score': 'False',
 'ccp_alpha': '0.0',
 'warm_start': 'False',
 'max_features': '1.0',
 'monotonic_cst': 'None',
 'n_jobs': 'None',
 'min_samples_split': '2'}

Answer: `min_samples_split` is **2**

## Q4. Launch the tracking server locally

Your task is to

* launch the tracking server on your local machine
* select a SQLite db for the backend store and a folder called `artifacts` for the artifacts store

You should keep the tracking server running to work on the next two exercises that use the server.

In addition to `backend-store-uri`, what else do you need to pass to properly configure the server?

* `default-artifact-root`
* `serve-artifacts`
* `artifacts-only`
* `artifacts-destination`


Answer: To configure MLflow to store artifacts on your local computer, use the **`--default-artifact-root`** option when starting the MLflow server.

## Q5. Tune the hyperparameters of the model

1. Modify the script `hpo.py` and make sure that the validation RMSE is logged to the tracking server for each run of the hyperparameter optimization

In [ ]:
!cat hpo.py

2. Run the script without passing any parameters

In [20]:
!python hpo.py

2025/05/22 02:04:16 INFO mlflow.tracking.fluent: Experiment with name 'random-forest-hyperopt' does not exist. Creating a new experiment.
🏃 View run aged-hare-553 at: http://127.0.0.1:5000/#/experiments/1/runs/e211097dd64c4fafbf069750c9d6aa88

🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1                    

🏃 View run dapper-perch-176 at: http://127.0.0.1:5000/#/experiments/1/runs/9e20bcab90794c65bf682bfda9c7f8f9

🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1                    

🏃 View run serious-hog-582 at: http://127.0.0.1:5000/#/experiments/1/runs/d5067c9ed4d34ca68eb9a2b020102687

🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1                    

🏃 View run sassy-kit-641 at: http://127.0.0.1:5000/#/experiments/1/runs/a82acf3a55714631b298f4403a42bb81

🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1                    

🏃 View run suave-ant-698 at: http://127.0.0.1:5000/#/experiments/1/runs/a8e86fb92e4c4ba3861868b5797a1ce3

🧪 V

3. What's the best validation RMSE that you got?

In [ ]:
# get tracking_uri from the script `hpo.py`
from mlflow.tracking import MlflowClient

client = MlflowClient(tracking_uri="http://127.0.0.1:5000")

In [30]:
client.search_experiments()

[<Experiment: artifact_location='/Users/nop/Documents/GitHub/mlops-zoomcamp-datatalks/02-experiment-tracking/artifacts/1', creation_time=1747854256919, experiment_id='1', last_update_time=1747854256919, lifecycle_stage='active', name='random-forest-hyperopt', tags={}>,
 <Experiment: artifact_location='/Users/nop/Documents/GitHub/mlops-zoomcamp-datatalks/02-experiment-tracking/artifacts/0', creation_time=1747853209465, experiment_id='0', last_update_time=1747853209465, lifecycle_stage='active', name='Default', tags={}>]

In [32]:
best_runs = client.search_runs(
    experiment_ids="1",
    filter_string="",
    max_results=3,
    order_by=["metrics.RMSE ASC"]
)

for run in best_runs:
    print(f"run id: {run.info.run_id}, RMSE: {run.data.metrics['RMSE']:.4f}")

run id: 5b6ccba24de247878697be85bf4e4074, RMSE: 5.3354
run id: 9cc8475537b248d8b1461388ab963be3, RMSE: 5.3547
run id: 06df9d8cd13e4f6c873d41f95e4001c4, RMSE: 5.3550


Answer: The best validation RMSE is **5.3354**

## Q6. Promote the best model to the model registry

1. Update the script `register_model.py` so that it selects the model with the lowest RMSE on the test set and registers it to the model registry

In [ ]:
!cat register_model.py

In [34]:
!python register_model.py

2025/05/22 02:25:50 INFO mlflow.tracking.fluent: Experiment with name 'random-forest-best-models' does not exist. Creating a new experiment.
🏃 View run gentle-ram-760 at: http://127.0.0.1:5000/#/experiments/2/runs/5cc2569493b6487ca82f19939f01e666
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/2
🏃 View run painted-dolphin-737 at: http://127.0.0.1:5000/#/experiments/2/runs/53627af8150e4f7ca50590e2f296b19d
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/2
🏃 View run amazing-mole-954 at: http://127.0.0.1:5000/#/experiments/2/runs/39e9f2d2ddad461da72008e8e478b04d
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/2
🏃 View run unleashed-foal-52 at: http://127.0.0.1:5000/#/experiments/2/runs/4acd40c7b6e2499b8c70b56f6dae0c2d
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/2
🏃 View run receptive-colt-729 at: http://127.0.0.1:5000/#/experiments/2/runs/69f7c5defb604107b4f95ea580342b97
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/2
Successfully re

2. What is the test RMSE of the best model?

In [35]:
run_reg_id = client.get_model_version(name="green_taxi_2023_duration_regressor", version="1").run_id
run_reg_v1 = client.get_run(run_id=run_reg_id)

round(run_reg_v1.data.metrics['test_rmse'], 3)

5.567

Answer: Test RMSE of the best model is **5.567**